crates/astria-sequencer/src/transaction/mod.rs

```diff
@@ -156,16 +160,98 @@ impl ActionHandler for Transaction {
+                Action::FeeChange(act) => act
+                    .check_authorization(state, from)
+                    .await
+                    .wrap_err("authorization check failed for FeeChangeAction")?,
+                Action::Ibc(_) => {
+                    // FIXME: this check should be moved. What's the correct
+                    // ibc AppHandler call to do it? Can we just update one of the trait methods
+                    // of crate::ibc::ics20_transfer::Ics20Transfer?
+                    ensure!(
+                        state
+                            .is_ibc_relayer(self)
+                            .await
+                            .wrap_err("failed to check if address is IBC relayer")?,
+                        "only IBC sudo address can execute IBC actions"
+                    );
+                }
+                Action::Ics20Withdrawal(act) => act
+                    .check_authorization(state, from)
+                    .await
+                    .wrap_err("authorization check failed for Ics20WithdrawalAction")?,
+                Action::IbcRelayerChange(act) => act
+                    .check_authorization(state, from)
+                    .await
+                    .wrap_err("authorization check failed for IbcRelayerChangeAction")?,
+                Action::FeeAssetChange(act) => act
+                    .check_authorization(state, from)
+                    .await
+                    .wrap_err("authorization check failed for FeeAssetChangeAction")?,
+                Action::InitBridgeAccount(act) => act
+                    .check_authorization(state, from)
+                    .await
+                    .wrap_err("authorization check failed for InitBridgeAccountAction")?,
+                Action::BridgeLock(act) => act
+                    .check_authorization(state, from)
+                    .await
+                    .wrap_err("authorization check failed for BridgeLockAction")?,
+                Action::BridgeUnlock(act) => act
+                    .check_authorization(state, from)
+                    .await
+                    .wrap_err("authorization check failed for BridgeUnlockAction")?,
+                Action::BridgeSudoChange(act) => act
+                    .check_authorization(state, from)
+                    .await
+                    .wrap_err("authorization check failed for BridgeSudoChangeAction")?,
+            }
+        }
+        Ok(())
+    }
+
     async fn check_and_execute<S: StateWrite>(&self, mut state: S) -> Result<()> {
         // Add the current signed transaction into the ephemeral state in case
         // downstream actions require access to it.
         // XXX: This must be deleted at the end of `check_stateful`.
         let mut transaction_context = state.put_transaction_context(self);
 
+        // check authorization
+        self.check_authorization(&state, self.address_bytes())
```

Comment:
The authorization checks are still being ran during execution here @SuperFluffy

Reply:
This is the wrong place to call it. The `ActionHandler for Transaction` trait delegates to its inner actions.

The idea is that checks and executions are atomic on a per-action level.

This is not atomic and on top of it surprising: why is authorization handled at the transaction but not the action level?

Reply:
Ah, ok I see what you mean by calling it in the wrong place. I'll refactor it